protocols/rendezvous/src/codec.rs

```diff
@@ -0,0 +1,624 @@
+pub enum Message {
+    Register(NewRegistration),
+    RegisterResponse(Result<Ttl, ErrorCode>),
+    Unregister(Namespace),
+    Discover {
+        namespace: Option<Namespace>,
+        cookie: Option<Cookie>,
+        limit: Option<Ttl>,
+    },
+    DiscoverResponse(Result<(Vec<Registration>, Cookie), ErrorCode>),
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct Namespace(String);
+
+impl Namespace {
+    /// Creates a new [`Namespace`] from a static string.
+    ///
+    /// This will panic if the namespace is too long. We accepting panicking in this case because we are enforcing a `static lifetime which means this value can only be a constant in the program and hence we hope the developer checked that it is of an acceptable length.
+    pub fn from_static(value: &'static str) -> Self {
+        if value.len() > 255 {
+            panic!("Namespace '{}' is too long!", value)
+        }
+
+        Namespace(value.to_owned())
+    }
+
+    pub fn new(value: String) -> Result<Self, NamespaceTooLong> {
+        if value.len() > 255 {
+            return Err(NamespaceTooLong);
+        }
+
+        Ok(Namespace(value))
+    }
+}
+
+impl From<Namespace> for String {
+    fn from(namespace: Namespace) -> Self {
+        namespace.0
+    }
+}
+
+impl fmt::Display for Namespace {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl PartialEq<str> for Namespace {
+    fn eq(&self, other: &str) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl PartialEq<Namespace> for str {
+    fn eq(&self, other: &Namespace) -> bool {
+        other.0.eq(self)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Namespace is too long")]
+pub struct NamespaceTooLong;
+
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+pub struct Cookie {
+    id: Uuid,
+    namespace: Option<Namespace>,
+}
+
+impl Cookie {
+    /// Construct a new [`Cookie`] for a given namespace.
+    ///
+    /// This cookie will only be valid for subsequent DISCOVER requests targeting the same namespace.
+    pub fn for_namespace(namespace: Namespace) -> Self {
+        Self {
+            id: Uuid::new_v4(),
+            namespace: Some(namespace),
+        }
+    }
+
+    /// Construct a new [`Cookie`] for a DISCOVER request that inquires about all namespaces.
+    pub fn for_all_namespaces() -> Self {
+        Self {
+            id: Uuid::new_v4(),
+            namespace: None,
+        }
+    }
+
+    pub fn into_wire_encoding(self) -> Vec<u8> {
+        let namespace = self.namespace.map(|ns| ns.0).unwrap_or_default();
+
+        let mut buffer = Vec::with_capacity(16 + namespace.len());
+        buffer.extend_from_slice(self.id.as_bytes());
+        buffer.extend_from_slice(namespace.as_bytes());
+
+        buffer
+    }
+
+    pub fn from_wire_encoding(mut bytes: Vec<u8>) -> Result<Self, InvalidCookie> {
+        // check length early to avoid panic during slicing
+        if bytes.len() < 16 {
```

Comment:
There isn't any. As far as I can see, cookies are opaque bytes to the client and their format is not specified. For us, the cookie is a `Uuid` (which is always 16 bytes), concatenated with the namespace (if provided during discover). Hence, we need to have at least 16 bytes here to avoid panics during slicing.

Reply:
Do I understand correctly that an implementation using a 15 byte cookie would thus not be compatible with this implementation?

Reply:
Ah, right, thanks. I somehow assumed other entities in a network to treat this in some other way than an opaque list of bytes.